### DISCUSSION TITLE
Strategies to avoid rate-limiting errors

### DISCUSSION THREAD

**Discussion by kevinschaul** [sec-edgar/sec-edgar#194] [Ideas]
Hello all, and thanks to the contributors for a great project. I've been using this with success for a few months.

Recently (in the past week or two), I started running into errors related to rate-limiting. I think I am not alone. (See #190 #192 #193 for potentially related issues.)

Here's the full html of one of these errors:

>  Your Request Originates from an Undeclared Automated Tool
> 
> To allow for equitable access to all users, SEC reserves the right to limit requests originating from undeclared automated tools. Your request has been identified as part of a network of automated tools outside of the acceptable policy and will be managed until action is taken to declare your traffic.
> 
> Please declare your traffic by updating your user agent to include company specific information.
> 
> For best practices on efficiently downloading information from SEC.gov, including the latest EDGAR filings, visit sec.gov/developer . You can also sign up for email updates on the SEC open data program, including best practices that make it more efficient to download data, and SEC.gov enhancements that may impact scripted downloading processes. For more information, contact opendata@sec.gov .
> 
> For more information, please see the SECâs Web Site Privacy and Security Policy . Thank you for your interest in the U.S. Securities and Exchange Commission. 

I've spent the morning poking around the code, and I have a few ideas to help us follow these best practices and avoid hitting SEC when not necessary. These are generally unformed thoughts, hence the GitHub discussion. I'm hoping contributors and users can chime in with more ideas. Let me know if it would be better to make these individual GitHub issues.

- **Expose a way to add a custom user-agent string, and set some default value**
I've got this implemented and am about to submit a PR for this.

- **Bypass unnecessary CIK lookups**
For my use-case, I have a list of CIKs I want to get filings for. Currently the code still runs through CIKLookup, which I suspect can be avoided. We could also consider storing these somehow (with an easy way for users to refresh them), as they shouldn't change too often.

- **Ensure rate-limiting applies to all requests**
Currently `NetworkClient`'s `rate_limit` argument only applies to the downloading/saving of files, but I am getting rate-limited on the CIK lookup portion of the code via calls to `get_response()`/`get_soup()`. If SEC strictly enforces a 10-requests/second policy, we should ensure all requests can follow it.

- **Ensure rate-limiting errors are properly raised**
For example, `NetworkClient` [attempts](https://github.com/sec-edgar/sec-edgar/blob/2c258a1d113d9e7c5cef498436709b68724c91ce/secedgar/client/network_client.py#L119-L123) to error out when rate-limited, but this code does not quite work. The error status code I receive is `403`. Further, this error is caught by an overbroad [try/catch](https://github.com/sec-edgar/sec-edgar/blob/2c258a1d113d9e7c5cef498436709b68724c91ce/secedgar/cik_lookup.py#L107-L113) in CIKLookup.

- **Find ways to test for these errors**
I'd be very curious if folks have ideas for how to test rate-limiting error-handling code. It seems quite tricky!

I am spending the day trying to get my specific use-case to work. I'll let you all know what I learn.

Anyone else have thoughts?

**Comment by jackmoody11** [comment]
Thanks for all of this @kevinschaul! The first two bullet points you have addressed here should be solved with #195 and #196 (thanks to you!) for the most part. While CIKLookup will call the `get_cik_map` function still in the changes proposed in #196 (as I have them currently), I don't think that this should make a big difference, as it is only one call and it typically takes 1/2 a second for me to run it. If speed is a big issue for you, then we can add caching, but I don't suspect this is the case. We could also add a check to see if all CIKs are digits in addition to checking if any of the CIKs are digits before looking them up via the SEC by combining your changes with mine.

As for ensuring that rate-limiting applies to all requests, I definitely agree. I am wondering if we can leverage `NetworkClient.fetch` to do this. This way, we can use throttling through aiohttp (no extra logic needed here) and get the benefits of async requests.

With regards to making sure that rate-limiting errors are properly raised, I wonder if #195 will fix this issue. I believe once we add a user agent, this should go back to a 429 error (since the 403 error is being raised on the first request that you make, since you don't currently have a user agent), but am not sure. You are right in pointing out that our try/catch is too broad in CIKLookup. We should probably just check for a specific type of query error. I need to see what that query error would look like. I will raise an issue for this. (EDIT: See #197)

As for testing rate-throttling, this has been a tough one for me to solve. I would welcome any advice/feedback from people who know more about this. I currently have a [very basic test](https://github.com/sec-edgar/sec-edgar/blob/2c258a1d113d9e7c5cef498436709b68724c91ce/secedgar/tests/client/test_network_client.py#L169) that checks to see if the throttling is working as I would expect it to, but I am sure there are other (probably better) ways to do this.

Thanks for all of the thought you put into this @kevinschaul! This is really helpful for everybody using the package. I would love to hear more from you about how you are using secedgar and how we can make it better for you/others.

**Reply by kevinschaul** [reply]
Thanks @jackmoody11! Regarding CIKLookup and `get_cik_map`, I agree that call single call should not make a big difference.

> We could also add a check to see if all CIKs are digits in addition to checking if any of the CIKs are digits before looking them up via the SEC by combining your changes with mine.

I think this is a great idea. Let me adjust that PR (#196) with this strategy in mind, and we can go from there.
